src/Microsoft.Health.Fhir.TemplateManagement/ArtifactProviders/BlobTemplateProvider.cs

```diff
@@ -0,0 +1,99 @@
+
+        private readonly IMemoryCache _templateCache;
+
+        private readonly string _blobTemplateCacheKey = "cached-blob-templates";
+
+        private readonly int _segmentSize = 100;
+
+        public BlobTemplateProvider(BlobContainerClient blobContainerClient, IMemoryCache templateCache)
+        {
+            _blobContainerClient = blobContainerClient;
+            _templateCache = templateCache;
+        }
+
+        public async Task<List<Dictionary<string, Template>>> GetTemplateCollectionAsync(CancellationToken cancellationToken = default)
+        {
+            // read templates from cache if available
+            if (_templateCache.TryGetValue(_blobTemplateCacheKey, out var templateCache))
+            {
+                return (List<Dictionary<string, Template>>)templateCache;
+            }
+
+            var templateLookup = new Dictionary<string, Template>();
+            var templateNames = await ListBlobsFlatListing(_blobContainerClient, _segmentSize);
+
+            foreach (var templateName in templateNames)
+            {
+                var template = await DownloadBlobToStringAsync(_blobContainerClient, templateName);
+                var renamedTemplate = TrimBlobName(templateName);
+                Template liquidTemplate = TemplateUtility.ParseTemplate(renamedTemplate, template);
+                templateLookup.Add(renamedTemplate, liquidTemplate);
+            }
+
+            var templates = new List<Dictionary<string, Template>>
+            {
+                templateLookup,
+            };
+
+            _templateCache.Set(_blobTemplateCacheKey, templates);
+
+            return templates;
+        }
+
+        private static async Task<List<string>> ListBlobsFlatListing(BlobContainerClient blobContainerClient, int? segmentSize)
+        {
+            var blobs = new List<string>();
+
+            var resultSegment = blobContainerClient.GetBlobsAsync()
+                .AsPages(default, segmentSize);
+
+            await foreach (Page<BlobItem> blobPage in resultSegment)
+            {
+                foreach (BlobItem blobItem in blobPage.Values)
+                {
+                    blobs.Add(blobItem.Name);
+                }
+            }
+
+            return blobs;
+        }
+
+        public static async Task<string> DownloadBlobToStringAsync(BlobContainerClient blobContainerClient, string blobName)
+        {
+            var blobClient = blobContainerClient.GetBlobClient(blobName);
+            BlobDownloadResult downloadResult = await blobClient.DownloadContentAsync();
+            return downloadResult.Content.ToString();
+        }
+
+        // TODO: eventually want to remove this and create a common utility for ACR and storage templates
+        private string TrimBlobName(string templateName)
+        {
+            return templateName
+                .Replace("Hl7v2/_", string.Empty)
+                .Replace("Hl7v2/", string.Empty)
```

Comment:
Should this include the folder prefix for other formats too like ccda, json, etc.?

Reply:
Yeah it should. I ended up switching to TemplateUtility.ParseTemplates which handles the filename parsing. I had been incorrectly using TemplateUtility.ParseTemplate previously which did not handle the filename parsing so this helper method was written but now that we've switched to TemplateUtility.ParseTemplates the method is no longer required so it has been removed.